# Wissensaggregator Mittelalter und frühe Neuzeit
WIAGweb2

## Daten für Domherren aus dem Digitalen Personenregister einlesen

Übertrage Daten aus der GS-online. Die Daten werden aus der produktiven Datenbank auf eine lokale Datenbank eingelesen: `gso`

[Literatur](#Literatur)  
[Domherren](#Domherren)
- [Einträge in `item`](#Einträge-in-item)
- [Daten eintragen](#Daten-eintragen)
- [Orden eintragen](#Orden-eintragen)
- [Literaturverweise eintragen](#Literaturverweise-eintragen)
- [Externe IDs eintragen](#Externe-IDs-eintragen)
- [Namensvarianten eintragen](#Namensvarianten-eintragen)
- [Suchvarianten eintragen (`name_lookup`)](#Suchvarianten-eintragen-(name_lookup))
- [Bezug zu den anderen Quellen (`canon_lookup`)](#Bezug-zu-den-anderen-Quellen)
- [Sortierung (`canon_sort`)](#canon_sort)  

[Nachtrag](#Nachtrag-person.note_name)

Quelle: https://mysql.gwdg.de  
*2022-04-27* Datenbank `gso_in_202204`  
*2022-05-12* Datenbank `gso_in_202205`  
*2022-06-02* Bischöfe und Domherren verweisen auf die jeweils aktuelle GSN  
*2022-07?* Datenbank `gso_in_202207`

Vorspann: Pakete laden; mit der Datenbank verbinden

In [13]:
gs_db = "gso_in_202210"

"gso_in_202210"

In [14]:
wds_path="../.."

"../.."

In [15]:
cd(wds_path)

In [16]:
using Pkg

In [17]:
Pkg.activate(".")

  Activating new project at `C:\Users\georg\Documents`


In [18]:
using Revise

In [19]:
using WiagDataSetup

In [20]:
using MySQL, DataFrames

In [21]:
Wds = WiagDataSetup

WiagDataSetup

In [22]:
Wds.setDBWIAG(user="georg", db="wiag2")

Passwort für User georg: ········


MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2")

Item Type ID

In [23]:
item_type_id = 6

6

In [24]:
ENV["COLUMNS"] = 120

120

In [25]:
data_path = "C:\\Users\\georg\\Documents\\projekte-doc\\WIAGweb2\\data"

"C:\\Users\\georg\\Documents\\projekte-doc\\WIAGweb2\\data"

In [26]:
out_path = "C:\\Users\\georg\\Documents\\projekte-doc\\WIAGweb2\\data_sql"

"C:\\Users\\georg\\Documents\\projekte-doc\\WIAGweb2\\data_sql"

## Löschungen für umfassendes Update

In [27]:
table_name = "name_lookup";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM name_lookup\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 16961, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [28]:
table_name = "familyname_variant";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM familyname_variant\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 2898, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [29]:
table_name = "givenname_variant";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM givenname_variant\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 1613, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [30]:
table_name = "id_external";
sql = "DELETE FROM $(table_name)
WHERE item_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM id_external\nWHERE item_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 10406, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [31]:
table_name = "person_role_property";
sql = "DELETE FROM $(table_name)
WHERE person_role_id IN
(SELECT id FROM person_role
WHERE person_id in (SELECT id from item WHERE item_type_id = $(item_type_id)))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM person_role_property\nWHERE person_role_id IN\n(SELECT id FROM person_role\nWHERE person_id in (SELECT id from item WHERE item_type_id = 6))", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [32]:
table_name = "person_role";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM person_role\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 16111, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [33]:
table_name = "person";
sql = "DELETE FROM $(table_name)
WHERE id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM person\nWHERE id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 9119, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [34]:
table_name = "item_reference";
sql = "DELETE FROM $(table_name)
WHERE item_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM item_reference\nWHERE item_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 7900, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

**Vorsicht** Zunächst die IDs (`id_public`) für die Domherren GS sichern, Einträge in `item` evtl. erst später löschen

In [100]:
table_name = "item";
sql = "DELETE FROM $(table_name)
WHERE item_type_id = $(item_type_id)";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM item\nWHERE item_type_id = 6", 0, -1, 8523, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

## Vorspann `gsn_gsn`

``` sql
use gso_in_202210;

CREATE TEMPORARY TABLE gsn_new AS
SELECT id, item_id, nummer
FROM gsn
WHERE id IN
(SELECT min(id) as id from gsn where not deleted group by item_id);

CREATE TABLE gsn_gsn AS
SELECT o.id, o.item_id, o.nummer, n.id as id_new, n.nummer as nummer_neu
FROM gsn AS o
INNER JOIN gsn_new AS n ON o.item_id = n.item_id
WHERE NOT o.deleted;
```

## Literatur

In [35]:
table_name = gs_db * ".books"
sql = "SELECT id, titel AS title, autoren AS author, nummer AS number_vol, " *
"uri AS online_resource, kurztitel AS shorttitle " *
"FROM $(table_name)";
df_vol = Wds.sql_df(sql);

In [36]:
names(df_vol)

6-element Vector{String}:
 "id"
 "title"
 "author"
 "number_vol"
 "online_resource"
 "shorttitle"

In [37]:
columns = [
    :id => :reference_id,
    :title => :full_citation,
    :author => :author_editor,
    :online_resource => :online_resource,
    :shorttitle => :title_short,
    :number_vol => :gs_volume_nr,
]

6-element Vector{Pair{Symbol, Symbol}}:
              :id => :reference_id
           :title => :full_citation
          :author => :author_editor
 :online_resource => :online_resource
      :shorttitle => :title_short
      :number_vol => :gs_volume_nr

In [38]:
df_vol_db = select(df_vol, columns);

In [39]:
df_vol_db[!, :item_type_id] .= item_type_id;

Bestehende Einträge löschen

In [40]:
table_name = "reference_volume"
sql = "DELETE FROM $(table_name) WHERE item_type_id = $(item_type_id)";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM reference_volume WHERE item_type_id = 6", 0, -1, 88, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

Daten eintragen

In [41]:
table_name = "reference_volume"
Wds.filltable!(table_name, df_vol_db, clear_table = false)

┌ Info: Rows inserted: 89
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


89

### Literatur GS
Feld `gs-citation`

In [42]:
using CSV, MySQL, DataFrames

In [43]:
filename = joinpath(data_path, "GS_Baende_ZitationGS.csv")

"C:\\Users\\georg\\Documents\\projekte-doc\\WIAGweb2\\data\\GS_Baende_ZitationGS.csv"

In [44]:
df_ref_c_gs = CSV.read(filename, DataFrame, delim="\t");

In [45]:
df_ref_c_gs[21:25, :]

5×2 DataFrame
 Row │ reference_id  gs_citation                       
     │ Int64         String                            
─────┼─────────────────────────────────────────────────
   1 │           21  Wendehorst, Alfred, Das Bistum W…
   2 │           22  Pauly, Ferdinand, Das Erzbistum …
   3 │           23  Maurer, Helmut, Das Bistum Konst…
   4 │           24  Sydow, Jürgen, Das Bistum Konsta…
   5 │           25  Kohl, Wilhelm, Das Bistum Münste…

In [46]:
table_name = "reference_volume"

"reference_volume"

In [47]:
sql = "UPDATE $(table_name) SET gs_citation = ? WHERE item_type_id = ? AND reference_id = ?";
stmt = DBInterface.prepare(Wds.dbwiag, sql);

In [26]:
DBInterface.execute(stmt, ("example_2", 6, 7))

MySQL.Cursor{true}(MySQL.API.MYSQL_STMT(Ptr{Nothing} @0x0000000086eac790), 0, Symbol[], Type[], Dict{Symbol, Int64}(), MySQL.API.BindHelper[], MySQL.API.MYSQL_BIND[], 0, 0, 0)

In [48]:
for row in eachrow(df_ref_c_gs)
    reference_id, gs_citation = row[[:reference_id, :gs_citation]]
    DBInterface.execute(stmt, (gs_citation, item_type_id, reference_id))
end
DBInterface.close!(stmt)    

## Domherren

Betrachte nur Personen, deren Eintrag nicht gelöscht ist und die online sind. Die maßgebliche GSN steht in dem Datensatz mit dem kleinsten Wert der id.
Ergebnisse *2022-02-25*
``` sql
use gso_in_202202;
select gsn.nummer, item_id, gsn.deleted, items.deleted, items.status
from gsn join items on gsn.item_id = items.id 
where items.status = 'online' and not items.deleted limit 12 offset 4000;
select count(*) from gsn join items on gsn.item_id = items.id where items.status = 'online' and not items.deleted;
```
98544
``` sql
select count(distinct(gsn.nummer)) 
from gsn join items on gsn.item_id = items.id 
where items.status = 'online' and not items.deleted;
```
98543
``` sql
create temporary table item_online as
select min(gsn.id) as gsn_tbl_id, gsn.nummer, item_id, items.deleted, items.status
from gsn join items on gsn.item_id = items.id where items.status = 'online'
group by items.id;
select count(*) from item_online;
```
87270

Finde Einträge, die von der Domherrendatenbank referenziert werden (sofern die verweisenden Domherren online sind!)

In [49]:
table_name = "item";
sql = "SELECT item.id as dh_item_id, e.value AS gsn
FROM wiag2.item 
JOIN wiag2.id_external AS e ON item.id = e.item_id
WHERE item_type_id = 5
AND authority_id = 200
AND item.is_online";
dh_gsn = Wds.sql_df(sql);

In [50]:
names(dh_gsn)

2-element Vector{String}:
 "dh_item_id"
 "gsn"

In [51]:
size(dh_gsn, 1)

1101

In [52]:
length(unique(dh_gsn.gsn))

1100

*2022-04-27* Zwei Domherren, die beide online sind, verweisen auf den gleichen Eintrag in der Personendatenbank. Die Redaktion weiß bescheid. Fahre nur mit einem der Einträge fort.  
*2022-05-12* die Werte stimmen überein: 606  
*2022-07-26*
*2022-10-15* die Werte stimmen nicht überein: 906, 905  
*2022-10-19* die Werte stimmen nicht überein: 1101, 1100

In [53]:
dh_gsn_g = groupby(dh_gsn, :gsn);

In [54]:
dh_gsn_gn = combine(dh_gsn_g, nrow);

In [55]:
subset(dh_gsn_gn, :nrow => ByRow(x -> x > 1))

1×2 DataFrame
 Row │ gsn            nrow  
     │ String?        Int64 
─────┼──────────────────────
   1 │ 036-00856-001      2

In [56]:
dh_gsn_f = combine(dh_gsn_g, first);

In [57]:
dh_gsn_f[51:55, :]

5×2 DataFrame
 Row │ gsn            dh_item_id 
     │ String?        Int32      
─────┼───────────────────────────
   1 │ 014-02050-001      451801
   2 │ 014-01488-001      446839
   3 │ 015-04122-001      453694
   4 │ 015-06070-001      453746
   5 │ 016-01329-001      453684

Lies aus `gsn` `item_id` (= item der Personendatenbank)

In [58]:
table_name = gs_db * ".gsn"
sql = "SELECT gsn.id as gsn_id, item_id, gsn.nummer, items.status 
FROM $(table_name)
JOIN $(gs_db).items ON item_id = items.id
AND NOT items.deleted AND items.status = 'online';"
item_all = Wds.sql_df(sql);

In [59]:
dh_item_gsn = innerjoin(dh_gsn_f, item_all, on = :gsn => :nummer);

In [60]:
size(dh_item_gsn, 1)

1100

In [61]:
setdiff(dh_gsn.gsn, item_all.nummer)

Union{Missing, String}[]

*2022-04-27* Diese GSN ist ungültig. Sie gehört zu Ägidius von der Lanken. Der Eintrag ist in einem anderen Eintrag aufgegangen.

Ermittle die IDs weiterer Domherren anhand der relevanten Stifte.

In [62]:
sql = "SELECT id_gsn FROM wiag2.institution AS inst
JOIN wiag2.item_property AS prp ON inst.id = prp.item_id
WHERE prp.name = 'domstift_short'"
df_dft = Wds.sql_df(sql);

In [63]:
ids_dft = "(" * join(df_dft.id_gsn, ", ") * ")"

"(3490, 3491, 3492, 3493, 3494, 3495, 3496, 3498, 3499, 3500, 3501, 3502, 3503, 2066, 953, 832, 783, 792, 794, 803, 736, 739, 675, 676, 679, 701, 616, 628, 343, 226, 3489, 3488, 3487, 60130)"

Betrachte nur gültige Amtseinträge

In [64]:
table_name = gs_db * ".offices";
sql = "SELECT DISTINCT(person_id), item_id 
FROM $(table_name) 
JOIN $(gs_db).persons AS p ON p.id = person_id
WHERE klosterid in $(ids_dft) AND NOT offices.deleted"
mon_item = Wds.sql_df(sql);

In [65]:
size(mon_item, 1)

9131

In [66]:
str_item = "(" * join(mon_item.item_id, ", ") * ")";

In [67]:
table_name = gs_db * ".gsn"
sql = "SELECT min(gsn.id) as id, item_id
FROM $(table_name)
JOIN $(gs_db).items ON item_id = items.id
WHERE item_id IN $(str_item)
AND NOT items.deleted AND items.status = 'online'
GROUP BY item_id";
mon_gsn_id = Wds.sql_df(sql);

In [68]:
size(mon_gsn_id, 1)

9065

In [69]:
str_gsn_id = "(" * join(mon_gsn_id.id, ", ") * ")";

In [70]:
table_name = gs_db * ".gsn"
sql = "SELECT gsn.id as gsn_id, item_id, gsn.nummer as gsn, items.status
FROM $(table_name)
JOIN $(gs_db).items ON item_id = items.id
WHERE gsn.id in $(str_gsn_id)";
mon_item_gsn = Wds.sql_df(sql);

In [71]:
size(mon_item_gsn, 1)

9065

Ohne die Einträge, die über die Domherrenbank referenziert werden.

In [72]:
mon_item_gsn_delta = antijoin(mon_item_gsn, dh_item_gsn, on = :item_id);

In [73]:
size(mon_item_gsn_delta)

(8192, 4)

In [74]:
item_gsn = vcat(mon_item_gsn_delta, select(dh_item_gsn, Not(:dh_item_id)));

In [75]:
size(item_gsn, 1)

9292

`persons`

In [76]:
table_name = gs_db * ".persons"
sql = "SELECT p.id as id_person_gso, item_id, " *
"namenspraefix, vorname, familienname, " *
"sterbedatum, geburtsdatum, `orden`, " *
"anmerkungen, titel, " * 
"gndnummer, viaf, " *
"familiennamenvarianten, vornamenvarianten, " *
"namenszusatz, " *
"zeitraum_von, zeitraum_bis " *
"FROM $(table_name) as p " *
"JOIN $gs_db.items ON items.id = p.item_id " *
"WHERE items.status = 'online' AND NOT items.deleted " *
"ORDER BY p.id";
df_p = Wds.sql_df(sql);

In [77]:
Wds.clean_up!(df_p)

In [78]:
one_valid(x, y) = !ismissing(x) | !ismissing(y)
subset!(df_p, [:vorname, :familienname] => ByRow(one_valid));

In [79]:
names(item_gsn)

4-element Vector{String}:
 "gsn_id"
 "item_id"
 "gsn"
 "status"

In [80]:
df_p_gsn = innerjoin(df_p, item_gsn, on = :item_id);

In [81]:
size(df_p_gsn, 1)

9292

### IDs zuordnen
Lies bestehende IDs.  


*2022-01-19* IDs sind von WIAG 1 zu übernehmen, später bei einem Update aus dem jeweils zuletzt verwendeten Datensatz (siehe unten)

In [56]:
sql = "SELECT id AS id_wiag, id_person_gso FROM wiag.cn_canon_gs"
df_id_cn_wiag = Wds.sql_df(sql);

LoadError: (1146): Table 'wiag.cn_canon_gs' doesn't exist

In [84]:
df_id_cn_wiag[201:205, :]

5×2 DataFrame
 Row │ id_public  id_person_gso 
     │ Int32      Int32?        
─────┼──────────────────────────
   1 │     80211           8556
   2 │     80212           8561
   3 │     80213           8606
   4 │     80214           8607
   5 │     80215           8620

*2022-04-27* IDs aus `wiag2_save`  
*2022-10-15* IDs aus `wiag2.item`  
*2022-10-19* IDs aus `wiag2.item`

In [82]:
db_save = "wiag2";
table_name = db_save * ".item";
sql = "SELECT id_public, id_in_source as id_person_gso
FROM $(table_name)
WHERE item_type_id = $(item_type_id)";
df_id_cn_wiag = Wds.sql_df(sql);

In [83]:
# nur wenn die Quelle wiag2 ist
function get_id_wiag(id_public)
    rgx = r"WIAG-Pers-CANON-([0-9]+)-0[0-9][0-9]"
    rgm = match(rgx, id_public)
    if isnothing(rgm)
        @warn "no match for" id_public
        return missing
    else
        return parse(Int, rgm[1])
    end
end

get_id_wiag (generic function with 1 method)

In [84]:
get_id_wiag("WIAG-Pers-CANON-80212-001")

80212

In [85]:
transform!(df_id_cn_wiag, :id_public => ByRow(get_id_wiag) => :id_wiag);

In [86]:
# nur wenn die Quelle wiag2 ist
transform!(df_p_gsn, :id_person_gso => ByRow(string) => :id_person_gso);

In [87]:
df_p_gsn[21:25, [:id_person_gso, :item_id, :vorname, :gsn, :status]]

5×5 DataFrame
 Row │ id_person_gso  item_id  vorname   gsn            status 
     │ String         Int32    String?   String?        String 
─────┼─────────────────────────────────────────────────────────
   1 │ 5049              5049  Matthäus  003-00513-001  online
   2 │ 5090              5090  Joachim   003-00554-001  online
   3 │ 5092              5092  Jakob     003-00556-001  online
   4 │ 5096              5096  Peter     003-00560-001  online
   5 │ 5135              5135  Gebhard   003-00599-001  online

In [88]:
df_id_cn_wiag[21:25, :]

5×3 DataFrame
 Row │ id_public                  id_person_gso  id_wiag 
     │ String?                    String?        Int64   
─────┼───────────────────────────────────────────────────
   1 │ WIAG-Pers-CANON-80010-001  4749             80010
   2 │ WIAG-Pers-CANON-80009-001  4719             80009
   3 │ WIAG-Pers-CANON-80008-001  4705             80008
   4 │ WIAG-Pers-CANON-80007-001  4698             80007
   5 │ WIAG-Pers-CANON-80006-001  4684             80006

In [89]:
df_cn_id = leftjoin(df_p_gsn, df_id_cn_wiag, on = :id_person_gso);

In [90]:
size(df_cn_id)

(9292, 22)

In [91]:
df_cn_id[201:205, [:id_person_gso, :item_id, :id_wiag, :vorname, :familienname]]

5×5 DataFrame
 Row │ id_person_gso  item_id  id_wiag  vorname    familienname 
     │ String         Int32    Int64?   String?    String?      
─────┼──────────────────────────────────────────────────────────
   1 │ 8561              8561    80212  Bruno      Querfurt
   2 │ 8606              8606    80213  Burchard   missing      
   3 │ 8607              8607    80214  Burchard   missing      
   4 │ 8620              8620    80215  Burchard   Belich
   5 │ 8623              8623    80216  Burchardt  Bornstedt

In [92]:
df_cn_id_id = subset(df_cn_id, :id_wiag => ByRow(!ismissing));
df_cn_id_new = subset(df_cn_id, :id_wiag => ByRow(ismissing));

In [93]:
size(df_cn_id_new, 1)

260

*2022-03-10* Es sind 83 neue IDs zu vergeben. (zunächst "nur" für WIAGweb2)  
*2022-04-27* Es sind 502 neue IDs zu vergeben.
*2022-05-12* Es sind 511 neue IDs zu vergeben.  
*2022-05-20* Es ist 1 neue IDs zu vergeben.  
*2022-07-26* Es sind 132 neue IDs zu vergeben.  
*2022-10-15* Es sind 51 neue IDs zu vergeben.  
*2022-10-19* Es sind 260 neue IDs zu vergeben.

In [94]:
next_id = maximum(df_cn_id_id.id_wiag) + 1

91039

In [95]:
sort!(df_cn_id_new, :item_id);

In [96]:
id = next_id
for row in eachrow(df_cn_id_new)
    row[:id_wiag] = id
    id += 1
end

In [97]:
sort!(df_cn_id_id, :id_wiag);

In [98]:
df_cn_id_id[end-5:end, [:item_id, :gsn_id, :id_wiag, :vorname, :familienname]]

6×5 DataFrame
 Row │ item_id  gsn_id  id_wiag  vorname      familienname 
     │ Int32    Int32   Int64?   String?      String?      
─────┼─────────────────────────────────────────────────────
   1 │  331538  372776    91033  Bruno        missing      
   2 │  331553  372853    91034  Heinrich I.  missing      
   3 │  331560  372877    91035  Hugo         missing      
   4 │  336969  380928    91036  Eberhard     missing      
   5 │  337014  381155    91037  Gebhard      Henneberg
   6 │  337237  381785    91038  Hunfried     missing      

In [99]:
df_cn_id_new[1:1, [:item_id, :gsn_id, :id_wiag, :vorname, :familienname]]

1×5 DataFrame
 Row │ item_id  gsn_id  id_wiag  vorname  familienname 
     │ Int32    Int32   Int64?   String?  String?      
─────┼─────────────────────────────────────────────────
   1 │   26661  281112    91039  Egilulf  missing      

In [100]:
df_cn_all = vcat(df_cn_id_id, df_cn_id_new);

In [101]:
size(df_cn_all, 1), size(unique(df_cn_all), 1)

(9292, 9292)

Otto mit `id_in_source` = "325778" ist doppelt. Wir nehmen ihn nur einmal auf. Die schon mal vergebene `id_public` wird nicht vermisst, wenn es schon immer einen Verweis von einem Domherren aus der Domherrendatenbank gab. Denn dann ist dessen `id_public` relevant.

In [102]:
unique!(df_cn_all);

### Einträge in `item`

entsprechende Einträge in Tabelle `item` einfügen

`id` für 'Domherr GS' in `item_type`: 6

In [103]:
df_cn_all[211:215, [:vorname, :familienname, :geburtsdatum, :sterbedatum, :id_wiag]]

5×5 DataFrame
 Row │ vorname     familienname  geburtsdatum  sterbedatum  id_wiag 
     │ String?     String?       String?       String?      Int64?  
─────┼──────────────────────────────────────────────────────────────
   1 │ Burchardus  missing       missing       missing        80225
   2 │ Busso       Bartensleben  missing       missing        80226
   3 │ Busso       Britzke       1534          1572           80232
   4 │ Busso       Schilt        missing       missing        80234
   5 │ Busso       Schraplau     missing       missing        80235

In [104]:
columns = [
    :id_person_gso => :id_in_source,
    :id_wiag => :id_wiag
]

2-element Vector{Pair{Symbol, Symbol}}:
 :id_person_gso => :id_in_source
       :id_wiag => :id_wiag

In [105]:
df_item = select(df_cn_all, columns);

In [106]:
using Dates

*2022-10-15 und später* user wiag hat die ID 40

In [107]:
user_wiag_id = 40;
time_stamp = Dates.format(now(), dateformat"yyyy-mm-dd HH:MM");

In [108]:
insertcols!(
    df_item, 
    1,
    :item_type_id => item_type_id,
    :edit_status => "importiert",
    :created_by => user_wiag_id,
    :date_created => time_stamp,
    :changed_by => user_wiag_id,
    :date_changed => time_stamp,
    :is_online => 1
);

In [109]:
df_item[17, :]

DataFrameRow
 Row │ item_type_id  edit_status  created_by  date_created      changed_by  date_changed      is_online  id_in_source  ⋯
     │ Int64         String       Int64       String            Int64       String            Int64      String        ⋯
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  17 │            6  importiert           40  2022-10-19 15:36          40  2022-10-19 15:36          1  4904          ⋯
                                                                                                        1 column omitted

Public ID erzeugen

In [110]:
function make_id_public(id::Real)
    id_public_key = "Pers-CANON"
    num_id_length = 5
    return "WIAG-" * id_public_key * "-" * lpad(id, num_id_length, '0') * "-001"
end

make_id_public (generic function with 1 method)

In [111]:
transform!(df_item, :id_wiag => ByRow(make_id_public) => :id_public);

In [112]:
names(df_item)

10-element Vector{String}:
 "item_type_id"
 "edit_status"
 "created_by"
 "date_created"
 "changed_by"
 "date_changed"
 "is_online"
 "id_in_source"
 "id_wiag"
 "id_public"

Bestehende Einträge löschen  
Die Einträge in `item` sind bei einem Update als letztes zu löschen, da die anderen Tabellen von Tabelle `item` abhängen.

In [113]:
item_type_id

6

In [114]:
table_name = "item";
sql = "DELETE FROM $(table_name) WHERE item_type_id = $(item_type_id)"
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM item WHERE item_type_id = 6", 0, -1, 9119, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [115]:
exclude_cols = [
    :id_wiag,    
]

1-element Vector{Symbol}:
 :id_wiag

In [116]:
table_name = "item";
Wds.filltable!(table_name, select(df_item, Not(exclude_cols)), clear_table = false)

┌ Info: Rows inserted: 9292
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


9292

| Datum | Zahl |
|:--- |:--- |
|*2022-04-27* | 8936 |
|*2022-07-26* | 9069 |
|*2022-10-15* | 9119 |
|*2022-10-19* | 9292 |

### Daten eintragen

IDs auslesen

In [117]:
table_name = "item";
sql = "SELECT id, id_in_source FROM $(table_name) where item_type_id = ($item_type_id)"
df_idx = DBInterface.execute(Wds.dbwiag, sql) |> DataFrame;

In [118]:
size(df_idx)

(9292, 2)

Spalte für die Zusammenführung befüllen (Typ: String!)

In [119]:
transform!(df_cn_all, :id_person_gso => ByRow(string) => :id_in_source);

In [120]:
df_cn_idx = innerjoin(df_idx, df_cn_all, on = :id_in_source);

In [121]:
length(unique(df_cn_idx.id_in_source)), size(df_cn_all, 1), size(df_cn_idx, 1)

(9292, 9292, 9292)

*2022-05-20* doppelter Verweis auf Otto in `item` bereinigt  
begin debug

In [173]:
function freq(v)
    f_count = Dict{eltype(v), Int}()
    for e in v
        f_count[e] = get(f_count, e, 0) + 1
    end
    return DataFrame(val = collect(keys(f_count)), freq = collect(values(f_count)))
end

freq (generic function with 1 method)

In [174]:
df_info_cn_n = freq(df_cn_idx.id_in_source);

In [175]:
subset(df_info_cn_n, :freq => ByRow((x -> x > 1)))

0×2 DataFrame

In [176]:
df_info_mult = subset(df_cn_idx, :id_in_source => ByRow(isequal("325778")))

1×24 DataFrame
 Row │ id      id_in_source  id_person_gso  item_id  namenspraefix  vorname  familienname  sterbedatum  geburtsdatum   ⋯
     │ Int32   String?       String         Int32    String?        String?  String?       String?      String?        ⋯
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ 308831  325778        325778          332194  missing        Otto     missing       wohl 1181    missing        ⋯
                                                                                                      15 columns omitted

In [120]:
df_info_mult[:, [:id_in_source, :vorname, :familienname, :gsn_id, :gsn]]

2×5 DataFrame
 Row │ id_in_source  vorname  familienname  gsn_id  gsn           
     │ String        String?  String?       Int32   String?       
─────┼────────────────────────────────────────────────────────────
   1 │ 325778        Otto     missing       374286  700-00388-001
   2 │ 325778        Otto     missing       374286  700-00388-001

In [124]:
df_cn_all_u = unique(df_cn_all);

In [125]:
size(df_cn_all_u)

(8957, 23)

*2022-05-12* Zwei Domherren verweisen über unterschiedliche GSN auf den gleichen Domherren GS
*2022-05-20* Die Domherren 13486 und 38149 verweisen beide auf 700-00388-001

end debug

Jahreszahlen ermitteln

In [122]:
parse_date_birth(s) = Wds.parsemaybe(s, :lower)

parse_date_birth (generic function with 1 method)

In [123]:
parse_date_death(s) = Wds.parsemaybe(s, :upper)

parse_date_death (generic function with 1 method)

In [124]:
transform!(df_cn_idx, :geburtsdatum => ByRow(parse_date_birth) => :num_date_birth);

In [125]:
transform!(df_cn_idx, :sterbedatum => ByRow(parse_date_death) => :num_date_death);

In [126]:
insertcols!(df_cn_idx,
    :item_type_id => item_type_id,
);

In [127]:
size(df_cn_idx)

(9292, 27)

Orden zuordnen

Bereinige Einträge

In [128]:
abbrev_dict = Dict("Kapuziner" => "OFMCap", "S. J." => "SJ")
get_abbreviation(order) = get(abbrev_dict, order, order)

get_abbreviation (generic function with 1 method)

In [129]:
transform!(df_cn_idx, :orden => ByRow(get_abbreviation) => :orden);

In [130]:
sql = "SELECT id as religious_order_id, abbreviation FROM religious_order";
df_rel_order = Wds.sql_df(sql);

In [131]:
df_cn_idx = leftjoin(df_cn_idx, df_rel_order, on = :orden => :abbreviation, matchmissing = :notequal);

In [132]:
a_not_b(a, b) = !ismissing(a) && ismissing(b)

a_not_b (generic function with 1 method)

In [133]:
df_cn_no_match = subset(df_cn_idx, [:orden, :religious_order_id] => ByRow(a_not_b));

In [134]:
df_cn_no_match[:, [:id, :id_in_source, :orden]]

0×3 DataFrame

In [135]:
columns = [
    :id => :id,
    :item_type_id => :item_type_id,
    :namenspraefix => :prefixname,
    :vorname => :givenname,
    :familienname => :familyname,
    :anmerkungen => :comment,
    :geburtsdatum => :date_birth,
    :sterbedatum => :date_death,
    :num_date_birth => :num_date_birth,
    :num_date_death => :num_date_death,
    :religious_order_id => :religious_order_id,
]

11-element Vector{Pair{Symbol, Symbol}}:
                 :id => :id
       :item_type_id => :item_type_id
      :namenspraefix => :prefixname
            :vorname => :givenname
       :familienname => :familyname
        :anmerkungen => :comment
       :geburtsdatum => :date_birth
        :sterbedatum => :date_death
     :num_date_birth => :num_date_birth
     :num_date_death => :num_date_death
 :religious_order_id => :religious_order_id

Einträge löschen (Korrektur/Update)

In [136]:
df_cn_db = select(df_cn_idx, columns);

In [137]:
table_name = "person"
sql = "DELETE FROM $(table_name)
WHERE id IN (SELECT id FROM item WHERE item_type_id = $(item_type_id))"
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM person\nWHERE id IN (SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [138]:
Wds.filltable!("person", df_cn_db; clear_table=false)

┌ Info: Rows inserted: 9292
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


9292

### Orden eintragen
*2022-01-19* Die (vorläufige) Ordenstabelle basiert auf den Ordensangaben für Bischöfe und auf den Ordensangaben hier.  
*2022-07-26, 2022-10-15* Orden sind schon zugeordnet (siehe oben)  

In [198]:
unique(df_cn_all.orden)

11-element Vector{Union{Missing, String}}:
 missing
 "OPraem"
 "SJ"
 "OP"
 "Kapuziner"
 "OFMObs"
 "S. J."
 "OSB"
 "OT"
 "CRVC"
 "OCist"

In [199]:
lookup_order = Dict(
    "Kapuziner" => "OFMCap",
    "S. J." => "SJ",
)

Dict{String, String} with 2 entries:
  "Kapuziner" => "OFMCap"
  "S. J."     => "SJ"

In [200]:
normalize_order(order) = get(lookup_order, order, order)

normalize_order (generic function with 1 method)

In [201]:
transform!(df_cn_idx, :orden => ByRow(normalize_order) => :orden);

In [202]:
orders_gs = filter(!ismissing, unique(df_cn_idx.orden))

9-element Vector{Union{Missing, String}}:
 "OPraem"
 "SJ"
 "OP"
 "OFMCap"
 "OFMObs"
 "OSB"
 "OT"
 "CRVC"
 "OCist"

In [156]:
table_name = "religious_order"
sql = "SELECT abbreviation FROM $(table_name)"
df_orders = Wds.sql_df(sql);

In [157]:
new_orders = setdiff(orders_gs, df_orders.abbreviation)

Union{Missing, String}[]

Eintrag in `item`

*2022-02-25* Nur nötig, wenn die Orden neu angelegt werden.

In [ ]:
table_name = "item";
sql = "SELECT * FROM $(table_name) LIMIT 0";
df_item = Wds.sql_df(sql);

In [ ]:
id_next = 22;

In [ ]:
item_type_order = 7;
user_wiag_id = 7; # Georg
time_stamp = Dates.format(now(), dateformat"yyyy-mm-dd HH:MM");

In [ ]:
df_item = select(df_item, Not(:id));

In [ ]:
table_name = "item"
sql = "INSERT INTO $(table_name) SET
    item_type_id = $(item_type_order),    
    id_in_source = $(id_next),
    is_deleted = 0,
    is_online = 0,
    created_by = $(user_wiag_id),
    date_created = '$(time_stamp)',
    changed_by = $(user_wiag_id),
    date_changed = '$(time_stamp)'"
DBInterface.execute(Wds.dbwiag, sql)

Lies `item_id` für religiöse Orden aus

In [144]:
table_name = "item"
sql = "SELECT id, id_in_source 
FROM $(table_name) 
WHERE id_in_source = $(id_next) and item_type_id = $(item_type_order)"
df_idx = Wds.sql_df(sql);

LoadError: UndefVarError: id_next not defined

In [145]:
item_id = df_idx[1, :id]

249263

In [146]:
table_name = "religious_order";
sql = "INSERT INTO $(table_name) SET id = $(item_id), abbreviation = 'CRVC'";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "INSERT INTO religious_order SET id = 249263, abbreviation = 'CRVC'", 0, -1, 1, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

### Literaturverweise eintragen

IDs wieder einlesen, falls hier neu gestartet wird.

In [139]:
table_name = "item";
sql = "SELECT id, id_in_source
FROM $(table_name)
WHERE item_type_id = $(item_type_id)";
df_idx = Wds.sql_df(sql);

In [140]:
size(df_idx, 1), size(df_p_gsn, 1)

(9292, 9292)

In [141]:
find_ids(id) = id in ("4586", "4590", "4620", "4623")

find_ids (generic function with 1 method)

In [142]:
subset(df_idx, :id_in_source => ByRow(find_ids))

4×2 DataFrame
 Row │ id      id_in_source 
     │ Int32   String?      
─────┼──────────────────────
   1 │ 480286  4623
   2 │ 480287  4620
   3 │ 480288  4590
   4 │ 480289  4586

In [143]:
df_p_gsn[1:5, [:id_person_gso, :vorname]]

5×2 DataFrame
 Row │ id_person_gso  vorname  
     │ String         String?  
─────┼─────────────────────────
   1 │ 4586           Nikolaus
   2 │ 4590           Johann
   3 │ 4620           Lorenz
   4 │ 4623           Werner
   5 │ 4678           Johannes

In [144]:
transform!(df_p_gsn, :id_person_gso => ByRow(string) => :id_in_source);

In [145]:
df_cn_idx = innerjoin(df_p_gsn, df_idx, on = :id_in_source);

In [146]:
size(df_cn_idx, 1)

9292

In [147]:
table_name = gs_db * ".locations";
sql = "SELECT item_id as item_id_gs, seiten, book_id  FROM $(table_name) " * 
"WHERE item_status = 'online' AND NOT deleted " *
"GROUP BY item_id, book_id, seiten"
df_p_ref = Wds.sql_df(sql);

Verbinde die Literaturangaben mit den relevanten Domherren

In [148]:
columns = [:id => :item_id, :item_id => :item_id_gs]

2-element Vector{Pair{Symbol, Symbol}}:
      :id => :item_id
 :item_id => :item_id_gs

In [149]:
df_p_ref = innerjoin(df_p_ref, select(df_cn_idx, columns), on = :item_id_gs);

In [150]:
size(df_p_ref)

(15480, 4)

Übernimm für Domherren mit Kurzbiographie nur Literaturhinweise mit Kurzbiographien (`transfer_gso.sql`: 2021-04-26) Es ist freilich möglich, dass es mehrere Literaturverweise gibt, die eine Kurzbiographie enthalten.

In [151]:
is_bio(seiten) = contains(seiten, "<b>");

In [152]:
df_p_ref_bio = subset(df_p_ref, :seiten =>ByRow(is_bio));

In [153]:
size(df_p_ref_bio)

(8114, 4)

Bilde die Differenzmenge in Bezug auf Domherren, nicht in Bezug auf Literaturverweise

In [154]:
is_item_bio(item_id) = item_id in unique(df_p_ref_bio.item_id)

is_item_bio (generic function with 1 method)

In [155]:
df_p_ref_mg_bio = subset(df_p_ref, :item_id => ByRow(!is_item_bio));

Einträge löschen für Korrekturen oder ein Update

In [156]:
item_type_id

6

In [157]:
table_name = "item_reference";
sql = "DELETE FROM $(table_name) WHERE item_type_id = $(item_type_id)";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM item_reference WHERE item_type_id = 6", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [158]:
insertcols!(df_p_ref_bio, 1, :item_type_id => item_type_id);

In [159]:
insertcols!(df_p_ref_mg_bio, 1, :item_type_id => item_type_id);

In [160]:
columns = [
    :item_id => :item_id,
    :seiten => :page,
    :item_type_id => :item_type_id,
    :book_id => :reference_id,
]

4-element Vector{Pair{Symbol, Symbol}}:
      :item_id => :item_id
       :seiten => :page
 :item_type_id => :item_type_id
      :book_id => :reference_id

In [161]:
table_name = "item_reference"
Wds.filltable!(table_name, select(df_p_ref_bio, columns))

┌ Info: Rows inserted: 8114
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


8114

*2022-02-25* Es werden nur noch Literaturangaben **mit Kurzbiographie** berücksichtigt.

In [ ]:
table_name = "item_reference"
Wds.filltable!(table_name, select(df_p_ref_mg_bio, columns))

Probe:
``` sql
select count(distinct(item_id)) from item_reference where item_type_id = 6;
```
7389

Es gibt also für jeden Domherren mindestens einen Literaturverweis

### Externe IDs eintragen

In [140]:
using MySQL, DataFrames

In [162]:
Wds.setDBWIAG(user = "georg", db = "wiag2")

Passwort für User georg: ········


MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2")

In [162]:
item_type_id = 6

6

Einträge löschen (für den Fall eines Fehlers oder für ein Update)

In [163]:
sql = "DELETE FROM id_external " *
"WHERE item_id IN (SELECT id FROM item WHERE item_type_id = $(item_type_id))"
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM id_external WHERE item_id IN (SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [164]:
sql = "SELECT id, url_name_formatter FROM authority " *
"WHERE id IN (1, 2, 3, 4, 5, 200)"
Wds.sql_df(sql)

6×2 DataFrame
 Row │ id     url_name_formatter                
     │ Int32  String?                           
─────┼──────────────────────────────────────────
   1 │     1  Gemeinsame Normdatei (GND) ID
   2 │     2  Wikidata
   3 │     3  Wikipedia-Artikel
   4 │     4  VIAF-ID
   5 │     5  WIAG-ID
   6 │   200  Personendatenbank der Germania S…

**VIAF**, dann GND, GSN

In [165]:
authority_id = 4 # siehe oben

4

In [166]:
col = :viaf;
df_p_auth = dropmissing(df_cn_idx, col);
insertcols!(df_p_auth, 1, :authority_id => authority_id);
transform!(df_p_auth, col => :value);

In [167]:
size(df_p_auth)

(2, 24)

In [168]:
df_p_auth

2×24 DataFrame
 Row │ authority_id  id_person_gso  item_id  namenspraefix  vorname           familienname           sterbedatum  gebu ⋯
     │ Int64         String         Int32    String?        String?           String?                String?      Stri ⋯
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │            4  325149          331572  von            Joseph Christian  Hohenlohe-Bartenstein  1817         1740 ⋯
   2 │            4  283124          283111  von            Bruno             Isenberg               1258         miss
                                                                                                      17 columns omitted

In [169]:
columns = [
    :id => :item_id,
    :authority_id => :authority_id,
    :value => :value,
]

3-element Vector{Pair{Symbol, Symbol}}:
           :id => :item_id
 :authority_id => :authority_id
        :value => :value

In [170]:
table_name = "id_external"
Wds.filltable!("id_external", select(df_p_auth, columns))

┌ Info: Rows inserted: 2
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


2

**GND**

In [171]:
authority_id = 1 # siehe oben

1

In [172]:
col = :gndnummer
df_p_auth = dropmissing(df_cn_idx, col);
insertcols!(df_p_auth, 1, :authority_id => authority_id);
transform!(df_p_auth, col => :value);

In [173]:
size(df_p_auth)

(1323, 24)

In [174]:
table_name = "id_external"
Wds.filltable!("id_external", select(df_p_auth, columns))

┌ Info: Rows inserted: 1323
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


1323

**GSN**

In [175]:
authority_id = 200 # siehe oben

200

In [176]:
col = :gsn
df_p_auth = dropmissing(df_cn_idx, col);
insertcols!(df_p_auth, 1, :authority_id => authority_id);
transform!(df_p_auth, col => :value);

In [177]:
table_name = "id_external"
Wds.filltable!("id_external", select(df_p_auth, columns))

┌ Info: Rows inserted: 9292
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


9292

### Namensvarianten eintragen

In [178]:
using MySQL, DataFrames

In [179]:
col = :familiennamenvarianten;
df_fnv = dropmissing(df_cn_idx, col);

In [180]:
size(df_fnv)

(1959, 22)

In [181]:
df_exp = Wds.expand_column(df_fnv, col, delim = ",");

In [182]:
size(df_exp)

(2936, 22)

In [183]:
df_exp[150:157, [:vorname, :familienname, col]]

8×3 DataFrame
 Row │ vorname   familienname  familiennamenvarianten 
     │ String?   String?       String                 
─────┼────────────────────────────────────────────────
   1 │ Liudolf   Werther       Insula
   2 │ Levin     Doring        Doringer
   3 │ Leonhard  Kalben        Calven
   4 │ Leonhard  Kalben        Calvis
   5 │ Konrad    Wellen        Welle
   6 │ Konrad    Oppen         Opin
   7 │ Konrad    Oppen         Uppin
   8 │ Konrad    Oppen         Upin

Spalte `lang` hat den Default-Wert 'de'.

In [184]:
columns = [
    :id => :person_id,
    :familiennamenvarianten => :name,
]

2-element Vector{Pair{Symbol, Symbol}}:
                     :id => :person_id
 :familiennamenvarianten => :name

Einträge löschen (Korrekturen, Daten-Update)

In [185]:
item_type_id

6

In [186]:
table_name = "familyname_variant";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM familyname_variant\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [187]:
table_name = "familyname_variant"
Wds.filltable!(table_name, select(df_exp, columns))

┌ Info: Rows inserted: 2936
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


2936

Vornamen

In [188]:
col = :vornamenvarianten;
df_gnv = dropmissing(df_cn_idx, col);

In [189]:
size(df_gnv)

(1236, 22)

In [190]:
df_exp = Wds.expand_column(df_gnv, col, delim = r", *");

In [191]:
size(df_exp)

(1630, 22)

In [192]:
df_exp[150:157, [:vorname, :familienname, col]]

8×3 DataFrame
 Row │ vorname    familienname  vornamenvarianten 
     │ String?    String?       String            
─────┼────────────────────────────────────────────
   1 │ Heinrich   Bork          Heinricus
   2 │ Heinrich   missing       Hinricus
   3 │ Heinrich   missing       Heinricus
   4 │ Heinrich   missing       Heinrik
   5 │ Hermannus  Mettinge      Hermann
   6 │ Hermann    Hövel         Hermannus
   7 │ Hermann    Hövel         Herimannus
   8 │ Hermann    Bevern        Herimannus

In [193]:
columns = [
    :id => :person_id,
    col => :name,
]

2-element Vector{Pair{Symbol, Symbol}}:
                :id => :person_id
 :vornamenvarianten => :name

Einträge löschen (Korrekturen, Daten-Update)

In [194]:
item_type_id

6

In [195]:
table_name = "givenname_variant";
sql = "DELETE FROM $(table_name)
WHERE person_id IN
(SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM givenname_variant\nWHERE person_id IN\n(SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [196]:
table_name = "givenname_variant"
Wds.filltable!(table_name, select(df_exp, columns))

┌ Info: Rows inserted: 1630
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


1630

### Suchvarianten eintragen (name_lookup)

In [197]:
columns = [    
    :vorname => :givenname,
    :vornamenvarianten => :givenname_variant,
    :namenspraefix => :prefix_name,
    :familienname => :familyname,
    :familiennamenvarianten => :familyname_variant,
];
lookup_columns = Dict(columns)

Dict{Symbol, Symbol} with 5 entries:
  :vorname                => :givenname
  :namenspraefix          => :prefix_name
  :familiennamenvarianten => :familyname_variant
  :familienname           => :familyname
  :vornamenvarianten      => :givenname_variant

In [198]:
rename!(df_cn_idx, lookup_columns);

In [199]:
df_nl = Wds.create_name_lookup(df_cn_idx); # schaut gut aus

In [200]:
names(df_nl)

3-element Vector{String}:
 "person_id"
 "gn_fn"
 "gn_prefix_fn"

In [201]:
size(unique(df_nl.person_id))

(9292,)

In [202]:
df_nl_empty = filter(:gn_fn => ismissing, df_nl);

In [203]:
size(df_nl_empty)

(0, 3)

In [204]:
df_nl[2301:2312, :]

12×3 DataFrame
 Row │ person_id  gn_fn                           gn_prefix_fn                      
     │ Int64      String?                         String?                           
─────┼──────────────────────────────────────────────────────────────────────────────
   1 │    481880  Johann Wischel Raesfeld         Johann Wischel von Raesfeld
   2 │    481880  Johann Raesfeld                 Johann von Raesfeld
   3 │    481881  Johann Franz Anton Raesfeld     Johann Franz Anton von Raesfeld
   4 │    481881  Johann Raesfeld                 Johann von Raesfeld
   5 │    481882  Goswin Raesfeld                 Goswin von Raesfeld
   6 │    481883  Goswin Raesfeld                 Goswin von Raesfeld
   7 │    481884  Radlef                          missing                           
   8 │    481885  Rabodo                          missing                           
   9 │    481886  R.                              missing                           
  10 │    481887  Wilhelm Plettenberg             Wilhelm von Plettenberg
  11 │    481888  Werner Anton Adolf Plettenberg  Werner Anton Adolf von Plettenbe…
  12 │    481888  Werner Plettenberg              Werner von Plettenberg

Entferne Einträge für diese Art von Items

In [205]:
table_name = "name_lookup"
sql = "DELETE FROM $(table_name) 
WHERE person_id IN (SELECT id FROM item WHERE item_type_id = $(item_type_id))";
DBInterface.execute(Wds.dbwiag, sql)

MySQL.TextCursor{true}(MySQL.Connection(host="127.0.0.1", user="georg", port="3306", db="wiag2"), "DELETE FROM name_lookup \nWHERE person_id IN (SELECT id FROM item WHERE item_type_id = 6)", 0, -1, 0, MySQL.API.MYSQL_RES(Ptr{Nothing} @0x0000000000000000), Symbol[], Type[], Dict{Symbol, Int64}(), 0, 1, false)

In [206]:
names(df_nl)

3-element Vector{String}:
 "person_id"
 "gn_fn"
 "gn_prefix_fn"

In [207]:
table_name = "name_lookup"
Wds.filltable!(table_name, df_nl)

┌ Info: 10000
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1193
┌ Info: Rows inserted: 17217
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


17217

## Bezug zu den anderen Quellen
Siehe **Variante** unten  
Zuerst Update der Domherrendatenbank durchführen

GS-EP und DH-EP: alle;  
GS-DH ohne GS-EP  
DH-DH ohne DH-EP  
GS-GS ohne GS-DH ohne GS-EP  
schließliche EP-EP (sofern EP als Ziel vorkommt)

In [ ]:
df_cn_lp = DataFrame(person_id = Int[], person_id_canon = Int[])

GS - EP

In [ ]:
authority_id = 200; # GSN
item_type_ep = 4; # Bischof
item_type_gs = 6; # Domherr GS
sql = "SELECT gs_gsn.id as person_id, ep_gsn.id as person_id_canon FROM
(SELECT i.id, ext.value as id_gsn
FROM item as i 
JOIN id_external AS ext ON i.id = ext.item_id
WHERE i.is_online 
AND i.item_type_id = $(item_type_gs)
AND ext.authority_id = $(authority_id)) AS gs_gsn
JOIN (SELECT i.id, ext.value as id_gsn
FROM item as i 
JOIN id_external AS ext ON i.id = ext.item_id
WHERE i.is_online 
AND i.item_type_id = $(item_type_ep)
AND ext.authority_id = $(authority_id)) AS ep_gsn 
ON ep_gsn.id_gsn = gs_gsn.id_gsn"
df_gs_ep = Wds.sql_df(sql);

In [ ]:
size(df_gs_ep)

In [ ]:
df_gs_ep[530:535, :]

In [ ]:
append!(df_cn_lp, df_gs_ep);

DH - EP

In [ ]:
authority_id = 5; # WIAG
item_type_ep = 4; # Bischof
item_type_dh = 5; # Domherr
sql = "SELECT dh.id AS person_id, ep.id AS person_id_canon
FROM (SELECT i.id, ext_dh.value AS id_ep_public
FROM item as i 
JOIN id_external as ext_dh ON i.id = ext_dh.item_id
WHERE ext_dh.authority_id = $(authority_id)
AND i.item_type_id = $(item_type_dh)
AND i.is_online) AS dh
JOIN (SELECT i.id, i.id_public
FROM item AS i
WHERE i.item_type_id = $(item_type_ep)) AS ep
ON ep.id_public = dh.id_ep_public";
df_dh_ep = Wds.sql_df(sql);

In [ ]:
size(df_dh_ep)

In [ ]:
df_dh_ep[50:57, :]

In [ ]:
append!(df_cn_lp, df_dh_ep);

GS - DH ohne GS - EP (nur Neuzugänge)

In [ ]:
authority_id = 200; # GSN
item_type_dh = 5; # Domherr
item_type_gs = 6; # Domherr GS
sql = "SELECT gs_gsn.id as person_id, dh_gsn.id as person_id_canon FROM
(SELECT i.id, ext.value as id_gsn
FROM item as i 
JOIN id_external AS ext ON i.id = ext.item_id
WHERE i.is_online 
AND i.item_type_id = $(item_type_gs)
AND ext.authority_id = $(authority_id)) AS gs_gsn
JOIN (SELECT i.id, ext.value as id_gsn
FROM item as i 
JOIN id_external AS ext ON i.id = ext.item_id
WHERE i.is_online 
AND i.item_type_id = $(item_type_dh)
AND ext.authority_id = $(authority_id)) AS dh_gsn 
ON dh_gsn.id_gsn = gs_gsn.id_gsn"
df_gs_dh = Wds.sql_df(sql);

In [ ]:
size(df_gs_dh)

In [ ]:
in_person_id(id) = id in df_cn_lp.person_id

In [ ]:
df_gs_dh = subset(df_gs_dh, :person_id => ByRow(!in_person_id));

In [ ]:
size(df_gs_dh)

In [ ]:
df_gs_dh[250:255, :]

In [ ]:
append!(df_cn_lp, df_gs_dh);

DH - DH und GS - GS ohne DH - EP ohne GS - DH ohne GS - EP: nur Neuzugänge

In [ ]:
table_name = "item";
item_type_dh = 5;
item_type_gs = 6;
sql = "SELECT id as person_id, id as person_id_canon 
FROM $(table_name) WHERE item_type_id in ($(item_type_dh), $(item_type_gs)) AND is_online";
df_cn_rx = Wds.sql_df(sql);

In [ ]:
size(df_cn_rx)

In [ ]:
df_cn_rx = subset(df_cn_rx, :person_id => ByRow(!in_person_id));

In [ ]:
size(df_cn_rx)

In [ ]:
append!(df_cn_lp, df_cn_rx);

In [ ]:
size(df_cn_lp)

EP - EP, sofern sie als Ziel vorkommen

In [ ]:
table_name = "item";
item_type_ep = 4;
sql = "SELECT id FROM $(table_name) WHERE item_type_id = $(item_type_ep)"
df_ids_ep = Wds.sql_df(sql);

In [ ]:
ids_ep_lp = intersect(df_ids_ep.id, df_cn_lp.person_id_canon);

In [ ]:
size(ids_ep_lp)

In [ ]:
df_ep_ep = DataFrame(person_id = ids_ep_lp, person_id_canon = ids_ep_lp);

In [ ]:
append!(df_cn_lp, df_ep_ep);

In [ ]:
size(df_cn_lp)

In [ ]:
table_name = "canon_lookup";
Wds.filltable!(table_name, df_cn_lp, clear_table = true)

### Bezug zu anderen Quellen (2)
Neue Struktur von `canon_lookup`: Ein Datensatz pro Domherr mit den Feldern

`person_id_name`: ID der Person für die Namensdaten (Bischof > Domherr > Domherr GS)  
`person_id_role`: ID der Person für Amtsdaten
`prio_role`: Priorisierung der Amtsdaten (Domherr > Domherr GS > Bischof)  

`person_id_role` ist eindeutig. Für jeden Wert von `person_id_name` gibt es immer auch einen Eintrag, wo `person_id_name` und `person_id_role` übereinstimmen.

Domherrendatenbank

IDs der Domherren aus der Domherrendatenbank und die Verweise auf Bischöfe und Domherren aus der Personendatenbank

In [208]:
table_name = "item";
item_type_domherr = 5;
sql = "SELECT i.id as dh, e1.value AS dh_gsn, e2.value AS dh_episc_id
FROM $(table_name) AS i
LEFT JOIN id_external AS e1 ON e1.item_id = i.id AND e1.authority_id = 200
LEFT JOIN id_external AS e2 ON e2.item_id = i.id AND e2.authority_id = 5
WHERE i.item_type_id = $(item_type_domherr) AND i.is_online";
df_dh = Wds.sql_df(sql);

In [209]:
size(df_dh, 1)

5526

In [210]:
names(df_dh)

3-element Vector{String}:
 "dh"
 "dh_gsn"
 "dh_episc_id"

*2022-04-28* Suche Verweise auf den gleichen Domherrn GS. Mitteilung an die Redaktion siehe oben.

In [211]:
v_dh_gsn = sort(collect(skipmissing(df_dh.dh_gsn))); v_dh_gsn_u = unique(v_dh_gsn);

In [212]:
length(v_dh_gsn), length(v_dh_gsn_u)

(1101, 1100)

In [213]:
find_by_id(id) = id in (200040, 200193);
df_info = subset(df_dh, :dh => ByRow(find_by_id))

0×3 DataFrame

Alle Bischöfe

In [214]:
table_name = "item";
item_type_bishop = 4;
sql = "SELECT i.id as ep, i.id_public, e1.value AS ep_gsn
FROM $(table_name) AS i
LEFT JOIN id_external AS e1 ON e1.item_id = i.id AND e1.authority_id = 200
WHERE i.item_type_id = $(item_type_bishop) AND i.is_online";
df_ep = Wds.sql_df(sql);

In [215]:
df_ep[11:15, :]

5×3 DataFrame
 Row │ ep      id_public                      ep_gsn  
     │ Int32   String?                        String? 
─────┼────────────────────────────────────────────────
   1 │ 475099  WIAG-Pers-EPISCGatz-21107-001  missing 
   2 │ 475100  WIAG-Pers-EPISCGatz-21106-001  missing 
   3 │ 475101  WIAG-Pers-EPISCGatz-21105-001  missing 
   4 │ 475102  WIAG-Pers-EPISCGatz-21103-001  missing 
   5 │ 475103  WIAG-Pers-EPISCGatz-21102-001  missing 

DH - EP

In [216]:
df_dh_ep = leftjoin(df_dh, 
    select(df_ep, [:ep => :ep1, :id_public => :id_public, :ep_gsn => :ep_gsn]), 
    on = :dh_episc_id => :id_public, 
    matchmissing = :notequal);

In [217]:
size(df_dh_ep, 1)

5526

In [218]:
names(df_dh_ep)

5-element Vector{String}:
 "dh"
 "dh_gsn"
 "dh_episc_id"
 "ep1"
 "ep_gsn"

In [219]:
count(!ismissing, df_dh_ep.dh_episc_id), count(!ismissing, df_dh_ep.ep1)

(402, 402)

begin debug  
*2022-05-12* fixed

In [285]:
is_is_not(a, b) = !ismissing(a) && ismissing(b);

In [288]:
subset(df_dh_ep, [:dh_episc_id, :ep1] => ByRow(is_is_not))

2×5 DataFrame
 Row │ dh      dh_gsn         dh_episc_id                        ep1      ep_gsn  
     │ Int32   String?        String?                            Int32?   String? 
─────┼────────────────────────────────────────────────────────────────────────────
   1 │ 241129  060-02416-001  https://wiag-stage.uni-goettinge…  missing  missing 
   2 │ 242124  700-00397-001  https://wiag-vocab.adw-goe.de/id…  missing  missing 

end debug

*2022-04-17* Schließlich werden alle Bischöfe gefunden, nachdem Einträge in `id_external` korrigiert wurden

Ergänze `dh_gsn` mit `ep_gsn`.  
Das ist nicht mehr nötig. Die GSN werden beim Einlesen der Daten schon ergänzt und in `id_external` eingetragen.

In [220]:
count(!ismissing, df_dh_ep.dh_gsn)

1101

In [221]:
update = 0
for row in eachrow(df_dh_ep)
    if ismissing(row[:dh_gsn]) && !ismissing(row[:ep_gsn])
        row[:dh_gsn] = row[:ep_gsn]
        update += 1
    end
end
update

0

In [222]:
count(!ismissing, df_dh_ep.dh_gsn)

1101

GS  
ID und GSN der Domherren aus der Personendatenbank

In [223]:
table_name = "item";
item_type_domherr_gs = 6;
sql = "SELECT i.id as gs, e1.value AS gs_gsn
FROM $(table_name) AS i
JOIN id_external AS e1 ON e1.item_id = i.id AND e1.authority_id = 200
WHERE i.item_type_id = $(item_type_domherr_gs) AND i.is_online";
df_gs = Wds.sql_df(sql);

In [224]:
replace(sql, '\n' => ' ')

"SELECT i.id as gs, e1.value AS gs_gsn FROM item AS i JOIN id_external AS e1 ON e1.item_id = i.id AND e1.authority_id = 200 WHERE i.item_type_id = 6 AND i.is_online"

In [225]:
size(df_gs, 1)

9292

GS - EP

In [226]:
df_gs_ep = leftjoin(df_gs, 
    select(df_ep, [:ep => :ep2, :ep_gsn => :ep_gsn]),
    on = :gs_gsn => :ep_gsn, 
    matchmissing = :notequal);

In [227]:
size(df_gs_ep, 1)

9292

In [228]:
count(ismissing, df_gs_ep.gs_gsn), count(!ismissing, df_gs_ep.ep2)

(0, 775)

In [229]:
df_gs_ep[201:205, :]

5×3 DataFrame
 Row │ gs      gs_gsn         ep2    
     │ Int32   String?        Int32? 
─────┼───────────────────────────────
   1 │ 484922  077-00339-001  478750
   2 │ 484923  062-00548-001  479587
   3 │ 484931  077-00859-001  478702
   4 │ 484937  077-00707-001  478974
   5 │ 484940  053-00080-001  478594

Die Domherren aus GS haben alle eine GSN, daher kann `df_gs_ep` ergänzt werden aus `df_cn_lp`.

In [230]:
names(df_gs_ep)

3-element Vector{String}:
 "gs"
 "gs_gsn"
 "ep2"

In [231]:
names(df_dh_ep)

5-element Vector{String}:
 "dh"
 "dh_gsn"
 "dh_episc_id"
 "ep1"
 "ep_gsn"

In [232]:
df_cn_lp = outerjoin(df_dh_ep, 
    df_gs_ep,
    on = :dh_gsn => :gs_gsn, 
    matchmissing = :equal);

In [233]:
# gesamt, davon zugeordnet über GSN, Zahl der GSN für Domherren
size(df_cn_lp, 1), size(df_dh_ep, 1) + size(df_gs_ep, 1) - size(df_cn_lp, 1), count(!ismissing, df_dh_ep.dh_gsn)

(13718, 1100, 1101)

Wenn nicht alle Domherren aus der Domherrendatenbank eine Zuordnung unter den Domherren aus der Personendatenbank finden, sind die Domherren aus der Personendatenbank neu einzulesen.  
*2022-07-26* `(12013, 739, 742)` Die Differenz wird vorerst toleriert.
*2022-10-15* `(12776, 905, 906)` Eine GSN wurde doppelt verwendet.  
*2022-10-19* `(13718, 1100, 1101)` Eine GSN wurde doppelt verwendet.  

In [234]:
df_cn_lp[21:25, :]

5×7 DataFrame
 Row │ dh      dh_gsn         dh_episc_id                    ep1     ep_gsn         gs      ep2    
     │ Int32?  String?        String?                        Int32?  String?        Int32?  Int32? 
─────┼─────────────────────────────────────────────────────────────────────────────────────────────
   1 │ 454467  068-01694-001  WIAG-Pers-EPISCGatz-20975-001  479837  068-01694-001  484423  479837
   2 │ 454561  052-01190-001  WIAG-Pers-EPISCGatz-20265-001  479287  052-01190-001  484426  479287
   3 │ 455449  052-01635-001  WIAG-Pers-EPISCGatz-20792-001  479760  052-01635-001  484427  479760
   4 │ 444749  058-02258-001  WIAG-Pers-EPISCGatz-10232-001  477621  058-02258-001  484486  477621
   5 │ 452172  063-00418-001  WIAG-Pers-EPISCGatz-02602-001  475760  063-00418-001  484614  475760

Führe Verweise auf Bischöfe im Feld `ep` zusammen. Übernimm den Verweis aus der Domherrendatenbank und - falls dieser nicht vorhanden ist - den Verweis aus der Personendatenbank.

In [235]:
df_cn_lp[!, :ep] .= df_cn_lp.ep1;

In [236]:
update = 0
for row in eachrow(df_cn_lp)
    if ismissing(row[:ep]) && !ismissing(row[:ep2])
        row[:ep] = row[:ep2]
        update += 1
    end
end
update

470

Bringe die Verweise in eine flache Struktur:  
Die Personendaten beziehen sich auf den einen der Einträge Bischof > Domherr > Domherr GS.  
Die Ämterdaten werden alle eingetragen, so vorhanden: Domherr > Domherr GS > Bischof.

In [237]:
df_cn_lp_db = DataFrame(person_id_name = Int[], person_id_role = Int[], prio_role = Int[])

0×3 DataFrame

In [238]:
for row in eachrow(df_cn_lp)
    prio = 1
    if !ismissing(row[:ep])
        name = row[:ep]
        if !ismissing(row[:dh])
            push!(df_cn_lp_db, (name, row[:dh], prio))
            prio += 1
        end
        if !ismissing(row[:gs])
            push!(df_cn_lp_db, (name, row[:gs], prio))
            prio += 1
        end        
        push!(df_cn_lp_db, (name, row[:ep], prio))
    elseif !ismissing(row[:dh])
        prio = 1
        name = row[:dh]
        push!(df_cn_lp_db, (name, row[:dh], prio))
        prio += 1
        if !ismissing(row[:gs])
            push!(df_cn_lp_db, (name, row[:gs], prio))
        end
    else
        prio = 1
        push!(df_cn_lp_db, (row[:gs], row[:gs], prio))
    end
end        

In [239]:
df_cn_lp_db[1021:1025, :]

5×3 DataFrame
 Row │ person_id_name  person_id_role  prio_role 
     │ Int64           Int64           Int64     
─────┼───────────────────────────────────────────
   1 │         455211          481197          2
   2 │         454643          454643          1
   3 │         454643          481199          2
   4 │         454595          454595          1
   5 │         454595          481201          2

In [240]:
count(ismissing, df_cn_lp_db.person_id_name)

0

In [241]:
table_name = "canon_lookup"
Wds.filltable!(table_name, df_cn_lp_db; clear_table = true)

┌ Info: 10000
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1193
┌ Info: Rows inserted: 15691
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


15691

Version mit paralleler Struktur für `canon_lookup`  
*2022-03-10* vermutlich obsolet

`id` ist `dh` oder `gs`

In [ ]:
df_cn_lp_gs[!, :id] = copy(df_cn_lp_gs.dh);

In [ ]:
for row in eachrow(df_cn_lp_gs)
    if ismissing(row[:id])
        if !ismissing(row[:gs])
            row[:id] = row[:gs]
        else
            @warn "missing dh and gs" row
        end
    end
end          

`person_id` ist `ep` oder `dh` oder `gs`

In [ ]:
df_cn_lp_gs[!, :person_id] = copy(df_cn_lp_gs.ep);

In [ ]:
for row in eachrow(df_cn_lp_gs)
    if ismissing(row[:person_id])
        if !ismissing(row[:dh])
            row[:person_id] = row[:dh]
        elseif !ismissing(row[:gs])
            row[:person_id] = row[:gs]
        else
            @warn "missing ep and dh and gs" row
        end
    end
end          

Führe die Verweise auf Bischöfe zusammen

In [ ]:
names(df_cn_lp_gs)

In [ ]:
df_cn_lp_gs[!, :ep] = copy(df_cn_lp_gs.ep1);

In [ ]:
for row in eachrow(df_cn_lp_gs)
    if !ismissing(row[:ep2])
        if !ismissing(row[:ep])
            if row[:ep] != row[:ep2]
                println(row)
            end
        else
            row[:ep] = row[:ep2]
        end
    end
end               

Keine Ausgabe -> Alle Verweise auf Bischöfe stimmen überein

In [ ]:
count(!ismissing, df_cn_lp_gs.ep1), count(!ismissing, df_cn_lp_gs.ep2), count(!ismissing, df_cn_lp_gs.ep)

In der Domherrendatenbank wurden alle Verweise recherchiert.

In [ ]:
columns = [
    :id => :id,
    :person_id => :person_id,
    :dh => :canon,
    :gs => :canon_gs,
    :ep => :bishop
]

In [ ]:
table_name = "canon_lookup"
Wds.filltable!(table_name, select(df_cn_lp_gs, columns))

### canon_sort
*2022-07-26* nicht mehr in Verwendung

In [43]:
sql = "SELECT person_id, role_name, min(name_short) AS domstift_sort
FROM person_role AS p
JOIN institution AS i ON p.institution_id = i.id
JOIN canon_lookup AS c ON c.person_id_role = p.person_id
WHERE c.prio_role = 1
AND i.item_type_id = 3
GROUP BY person_id";
df_sort_i = Wds.sql_df(sql);

In [44]:
df_cn_lp_db = leftjoin(df_cn_lp_db, df_sort_i, on = :person_id_role => :person_id);

In [45]:
size(df_cn_lp_db)

(12952, 5)

In [64]:
sql = "SELECT person_id, min(date_sort_key) AS date_sort_key
FROM person_role AS p
JOIN canon_lookup AS c ON c.person_id_role = p.person_id
WHERE c.prio_role = 1
GROUP BY person_id";
df_sort_d = Wds.sql_df(sql);

In [65]:
df_cn_lp_db = leftjoin(df_cn_lp_db, df_sort_d, on = :person_id_role => :person_id);

In [66]:
df_cn_g = groupby(df_cn_lp_db, :person_id_name);

In [67]:
df_cn_s = combine(df_cn_g, first);

In [69]:
size(df_cn_s)

(11586, 6)

In [70]:
columns = [
    :person_id_name => :id,
    :domstift_sort => :domstift_short,
    :date_sort_key => :date_sort_key
]

3-element Vector{Pair{Symbol, Symbol}}:
 :person_id_name => :id
  :domstift_sort => :domstift_short
  :date_sort_key => :date_sort_key

In [71]:
table_name = "canon_sort"
Wds.filltable!(table_name, select(df_cn_s, columns); clear_table = true)

┌ Info: 10000
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1193
┌ Info: Rows inserted: 11586
└ @ WiagDataSetup C:\Users\georg\Documents\projekte\WiagDataSetup.jl\src\WiagDataSetup.jl:1216


11586

### Nachtrag person.note_name

In [242]:
using DataFrames

In [243]:
table_name = gs_db * "." * "persons";
sql = "SELECT id, vorname, familienname, namenszusatz
FROM $(table_name)
WHERE namenszusatz IS NOT NULL and namenszusatz <> ''";

In [244]:
sql_db = join(split(sql, "\n"), ' ')

"SELECT id, vorname, familienname, namenszusatz FROM gso_in_202210.persons WHERE namenszusatz IS NOT NULL and namenszusatz <> ''"

In [245]:
df_p_note_gs = Wds.sql_df(sql);

In [246]:
Wds.clean_up!(df_p_note_gs);

In [247]:
transform!(df_p_note_gs, :id => ByRow(string) => :id_in_source);

In [248]:
table_name = "item";
sql = "SELECT id as item_id, id_in_source, item_type_id, id_public
FROM $(table_name)
WHERE item_type_id = $(item_type_id)";
df_item = Wds.sql_df(sql);

In [249]:
size(df_item, 1)

9292

In [250]:
df_p_note = innerjoin(df_p_note_gs, df_item, on = :id_in_source);

In [251]:
size(df_p_note, 1)

771

Nur ein Teil der Personen aus GS sind Domherren.

In [252]:
df_p_note[200:205, [:id, :id_in_source, :item_id, :id_public, :vorname, :familienname]]

6×6 DataFrame
 Row │ id      id_in_source  item_id  id_public                  vorname   familienname 
     │ Int32   String        Int32    String?                    String?   String?      
─────┼──────────────────────────────────────────────────────────────────────────────────
   1 │ 108059  108059         482800  WIAG-Pers-CANON-82891-001  Johannes  Lepper
   2 │ 108054  108054         482802  WIAG-Pers-CANON-82889-001  Johannes  Jacobi
   3 │ 108042  108042         482807  WIAG-Pers-CANON-82884-001  Johannes  Hodelen
   4 │ 108039  108039         482808  WIAG-Pers-CANON-82882-001  Johannes  Hinxt
   5 │ 108038  108038         482809  WIAG-Pers-CANON-82881-001  Johannes  Hertoge
   6 │ 108157  108157         482812  WIAG-Pers-CANON-82941-001  Johannes  Werne

In [253]:
df_p_note[200:205, [:id, :id_in_source, :item_id, :namenszusatz, :vorname, :familienname]]

6×6 DataFrame
 Row │ id      id_in_source  item_id  namenszusatz           vorname   familienname 
     │ Int32   String        Int32    String                 String?   String?      
─────┼──────────────────────────────────────────────────────────────────────────────
   1 │ 108059  108059         482800  von Ahlen, der Ältere  Johannes  Lepper
   2 │ 108054  108054         482802  von Borken             Johannes  Jacobi
   3 │ 108042  108042         482807  von Dülmen             Johannes  Hodelen
   4 │ 108039  108039         482808  von Hamm               Johannes  Hinxt
   5 │ 108038  108038         482809  von Kamen              Johannes  Hertoge
   6 │ 108157  108157         482812  der Jüngere            Johannes  Werne

In [254]:
names(df_p_note)

8-element Vector{String}:
 "id"
 "vorname"
 "familienname"
 "namenszusatz"
 "id_in_source"
 "item_id"
 "item_type_id"
 "id_public"

In [255]:
table_name = "person";
function sql_update(note_name, person_id) 
    sql = "UPDATE $(table_name) AS p " *
    "SET p.note_name = '$(note_name)' " *
    "WHERE p.id = $(person_id)"
    return sql
end

sql_update (generic function with 1 method)

In [256]:
count_loop = 0;
count_msg = 200;
for row in eachrow(df_p_note)
    sql = sql_update(row[:namenszusatz], row[:item_id])
    DBInterface.execute(Wds.dbwiag, sql)
    count_loop += 1;
    if count_loop % count_msg == 0
        @info count_loop
    end
end

┌ Info: 200
└ @ Main In[256]:8
┌ Info: 400
└ @ Main In[256]:8
┌ Info: 600
└ @ Main In[256]:8
